docs: updating documentation for azureml (#6131)
mlflow/mlflow: docs/source/models.rst

.. _azureml_deployment:

Deploy a ``python_function`` model on Azure Machine Learning
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The MLflow plugin `azureml-mlflow <https://pypi.org/project/azureml-mlflow/>`_ can deploy models to Azure ML for real-time inference with Azure Kubernetes Service (AKS), Azure Container Instances (ACI) and Azure Machine Learning Managed Inference.

.. note:: 

>   Azure Machine Learning supports the deployment of MLflow models for batch-inference. However, this deployment target is not supported using the `azureml-mlflow <https://pypi.org/project/azureml-mlflow/>`_. Please use any of the `other supported deployment methods <https://docs.microsoft.com/en-us/azure/machine-learning/how-to-deploy-mlflow-models>`_.

Reviewer: ```suggestion
  Azure Machine Learning supports the deployment of MLflow models for batch inference. However, the `azureml-mlflow <https://pypi.org/project/azureml-mlflow/>`_ library does not provide batch inference functionality. If you wish to perform batch inference, please see https://docs.microsoft.com/en-us/azure/machine-learning/how-to-deploy-mlflow-models> for more information.
```

.. note:: 
  Azure Machine Learning supports the deployment of MLflow models for batch-inference. However, this deployment target is not supported using the `azureml-mlflow <https://pypi.org/project/azureml-mlflow/>`_. Please use any of the `other supported deployment methods <https://docs.microsoft.com/en-us/azure/machine-learning/how-to-deploy-mlflow-models>`_.

The resulting deployment accepts the following data formats as input:

* JSON-serialized pandas DataFrames in the ``split`` orientation. For example, ``data = pandas_df.to_json(orient='split')``. This format is specified using a ``Content-Type`` request header value of ``application/json``.
* Tensor input format as JSON-serialized lists (for tensors) and dictionary of lists (for named tensors).

.. warning::
    The ``TensorSpec`` input format is not fully supported for deployments on Azure Machine Learning at the moment. Be aware that many ``autolog()`` implementations may use ``TensorSpec`` for model's signatures when logging models and hence those deployments will fail in Azure ML.

> Deployments can be generated using both the Python API or MLflow CLI. In both cases, a ``JSON`` configuration file can be indicated with the details of the deployment you want to achieve. If not indicated, then a default deployment is done using Azure Container Instances (ACI) and a minimal configuration. The full specification of this configuration file can be checked at `Deployment configuration schema <https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema>`_. Also, you will also need the Azure ML MLflow Tracking URI of your particular Azure ML Workspace where you want to deploy your model. You can obtain this URI in several ways:

Reviewer: Is the following no longer accurate / relevant?

```
The full specification of this configuration file can be checked at `Deployment configuration schema <https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema>`_ 
```

Author: It was moved to another section. The JSON specification depends on the target used. AKS and ACI uses the same specification but Managed Endpoints uses a different one as it's a compute belonging to our V2 platform.

> .. note:: 

Reviewer 2: Can an example be added to this clarification that shows exactly what that JSON payload should look like (i.e. casting from a pd.DataFrame is probably sufficient) and how to set it within the SDK's API?

Author: I included a sample of the payload later. Are you suggesting something else?

Reviewer: @santiagxf The text in this paragraph refers to "this payload is different from the one used when models are served using `mlflow models serve`", which is confusing because the paragraph doesn't contain an example payload. Can we either move the payload or edit the text?

Author: Thank you @dbczumar. I think it is better to raise the reader's attention earlier knowing that the payload is different. The example is later in the documentation so we don't change the subject. They will continue reading for an example if relevant to them. I rather keep the notice here.

>   Regardless of the format, the input data should be indicated in a JSON payload within the key `input_data`. Notice that this payload is different from the one used when models are served using `mlflow models serve`.

Reviewer: ```suggestion
  Regardless of the format, the input data should be supplied as a JSON payload within the key `input_data`. Notice that this payload is different from the one used when models are served using `mlflow models serve`.
```

>   Regardless of the format, the input data should be indicated in a JSON payload within the key `input_data`. Notice that this payload is different from the one used for instance when models are served using `mlflow models serve`.

Reviewer 2: ```suggestion
  Regardless of the format, the input data should be indicated in a JSON payload within the key `input_data`. Notice that this payload is different from the one used when models are served using `mlflow models serve`.
```

  * Click "View all properties in Azure Portal" on the pane popup.
  * Copy the ``MLflow tracking URI`` value from the properties section.

> * Programmatically, using Azure ML SDK with the method `Workspace.get_mlflow_tracking_uri() <https://docs.microsoft.com/en-us/python/api/azureml-core/azureml.core.workspace.workspace?view=azure-ml-py#azureml-core-workspace-workspace-get-mlflow-tracking-uri>`_. If you are running inside Azure ML Compute (e.g., a Compute Instance) you this value can be accessed from the environment variable ``os.environ["MLFLOW_TRACKING_URI"]``.

Reviewer: ```suggestion
* Programmatically, using Azure ML SDK with the method `Workspace.get_mlflow_tracking_uri() <https://docs.microsoft.com/en-us/python/api/azureml-core/azureml.core.workspace.workspace?view=azure-ml-py#azureml-core-workspace-workspace-get-mlflow-tracking-uri>`_. If you are running inside Azure ML Compute (e.g., a Compute Instance), this value can be accessed from the environment variable ``os.environ["MLFLOW_TRACKING_URI"]``.
```

> * Programmatically, using Azure ML SDK with the method `Woskspace.get_mlflow_tracking_uri() <https://docs.microsoft.com/en-us/python/api/azureml-core/azureml.core.workspace.workspace?view=azure-ml-py#azureml-core-workspace-workspace-get-mlflow-tracking-uri>`_. If you are running inside Azure ML Compute, like for instance a Compute Instace, you can get this value also from the environment variable ``os.environ["MLFLOW_TRACKING_URI"]``.

Reviewer 2: ```suggestion
* Programmatically, using Azure ML SDK with the method `Workspace.get_mlflow_tracking_uri() <https://docs.microsoft.com/en-us/python/api/azureml-core/azureml.core.workspace.workspace?view=azure-ml-py#azureml-core-workspace-workspace-get-mlflow-tracking-uri>`_. If you are running inside Azure ML Compute (e.g., a Compute Instance) you this value can be accessed from the environment variable ``os.environ["MLFLOW_TRACKING_URI"]``.
```

Remarks:
 * If ``containerResourceRequirements`` is not indicated, a deployment with minimal compute configuration is applied (``cpu: 0.1`` and ``memory: 0.5``).
 * If ``location`` is not indicated, it defaults to the location of the workspace.

>  * The full specification of this configuration file can be viewed at `Deployment configuration schema <https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema>`_.

Reviewer: ```suggestion
 * The full specification of the deployment configuration schema can be viewed at https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema.
```

>  * The full specification of this configuration file can be checked at `Deployment configuration schema <https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema>`_.

Reviewer 2: ```suggestion
 * The full specification of this configuration file can be viewed at `Deployment configuration schema <https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema>`_.
```

    }

Remarks:

>   * In above exmaple, ``aks-mlflow`` is the name of an Azure Kubernetes Cluster registered/created in Azure Machine Learning.

Reviewer 2: ```suggestion
  * In the above example, ``aks-mlflow`` is the name of an Azure Kubernetes Cluster (AKS) 
 that has been registered and created in the Azure Machine Learning (AML) service.
```

  * The full specification of this configuration file can be checked at `Deployment configuration schema <https://docs.microsoft.com/en-us/azure/machine-learning/reference-azure-machine-learning-cli#deployment-configuration-schema>`_.

.. rubric:: Configuration example for an Managed Endpoint deployment

.. code-block:: json

    {
      "instance_type": "Standard_DS2_v2",
      "instance_count": 1,
    }

Remarks:

>   * In above exmaple, ``aks-mlflow`` is the name of an Azure Kubernetes Cluster registered/created in Azure Machine Learning.

Reviewer 2: ```suggestion
  * In above example, ``aks-mlflow`` is the name of an Azure Kubernetes Cluster (AKS) 
 service that has been registered and created in Azure Machine Learning (AML).
```

>   * The full specification of this configuration file can be checked at `Managed online deployment schema (v2) <https://review.docs.microsoft.com/en-us/azure/machine-learning/reference-yaml-deployment-managed-online>`_.

Reviewer 2: ```suggestion
  * The full specification of this configuration file can be viewed at `Managed online deployment schema (v2) <https://review.docs.microsoft.com/en-us/azure/machine-learning/reference-yaml-deployment-managed-online>`_.
```

>   * The full specification of this configuration file can be viewed at `Managed online deployment schema (v2) <https://review.docs.microsoft.com/en-us/azure/machine-learning/reference-yaml-deployment-managed-online>`_.

Reviewer: ```suggestion
 * The full specification of the AKS deployment configuration schema (v2) can be viewed at https://review.docs.microsoft.com/en-us/azure/machine-learning/reference-yaml-deployment-managed-online.
```

    import json
    # `sample_input` is a JSON-serialized pandas DataFrame with the `split` orientation
    sample_input = {

>         "input_data": {

Reviewer 2: This is a breaking change, right? 
Can you provide a version context of when this change takes place (version number of the SDK) and provide a note on a lack of backwards compatibility (or, if backwards compatibility is handled, provide some indication that this is the case)?

Author: I think this functionality has been incorrectly documented here, cause this is the way it has worked from the beginning. The key `input_data` has always been required. I'm updating this documentation for that reason. I also added the note about this payload being different from the one MLflow uses. Do you think we should add something else?

Author: @BenWilson2  let me know what's your opinion around this.

Reviewer: Thanks for clarifying / correcting the docs here!
